pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -398,3 +406,55 @@ def _create_remote_artifact(d_artifact, content_artifact):
         )
         ra.validate_checksums()
         return ra
+
+
+class ACSArtifactHandler(Stage):
+    """
+    A Stages API stage to download :class:`~pulpcore.plugin.models.Artifact` files from Alternative
+    Content Source if available.
+    """
+
+    async def run(self):
+        async for batch in self.batches():
+            if AlternateContentSource.objects.exists():
+                # get first checksum type from d_artifacts to use for search ACS
+                # or fallback for sha256
+                checksum_type = "sha256"
+                for d_content in batch:
+                    for d_artifact in d_content.d_artifacts:
+                        for cks_type in d_artifact.artifact.COMMON_DIGEST_FIELDS:
+                            if getattr(d_artifact.artifact, cks_type):
+                                checksum_type = cks_type
+                                break
+                        else:
+                            continue
+                        break
+                    else:
+                        continue
+                    break
+
+                batch_remote_artifacts_checksums = [
+                    getattr(d_artifact.artifact, checksum_type)
+                    for d_content in batch
+                    for d_artifact in d_content.d_artifacts
+                ]
+                acs_repos = Repository.objects.filter(
+                    alternatecontentsourcepath__in=AlternateContentSourcePath.objects.all()
+                )
+                existing_ras = RemoteArtifact.objects.filter(
+                    content_artifact__content__repositories__in=acs_repos,
+                    **{checksum_type + "__in": batch_remote_artifacts_checksums},
+                ).select_related("remote")
```

Comment:
I think you could probably simplify this using [the `acs()` queryset method](https://github.com/pulp/pulpcore/pull/1605/files#diff-81f6a78175bb93934b6beff952646d3ca1ef3731f1ff14492d4ec77bfc3fdf82R574-R576):

```suggestion
                existing_ras = RemoteArtifact.objects.acs().filter(
                    **{f"{checksum_type}__in": batch_remote_artifacts_checksums}
                )
```